### DISCUSSION TITLE
Is it possible to set up dynamic error messages based on validation rules?

### DISCUSSION THREAD

**Discussion by jvgeee** [react-hook-form/react-hook-form#7675] [Q&A] [answered]
I'm using controlled inputs in React Native, and I'd like to set up validation rules for my form that give different error messages based on the current state.

Case 1: User is typing / field is active, I want to show no error messages
Case 2: Field is blurred, I want to run a validation function

If length === 0, show "You must enter this field"
If length === 1, show "Field must be at least 2 characters"
If length >= 2, it's valid

I'm currently trying to do this by passing this function to `rules`:

```
// Validation handler
export const handleValidationErrors = (minLength = 1, maxLength?, messages?) => (item) => {
  const { length = 0 } = item;
  const { defaultMsg = "This field is required" } = messages || {};

  const isValid = maxLength ? length >= minLength && length <= maxLength : length >= minLength;

  if (isValid) return false;

  if (messages && messages[length]) return messages[length];

  return defaultMsg;
};

// In usage
const {
    field: { onChange, value },
    fieldState: { invalid, isDirty },
  } = useController({
    name,
    control: formControls.control,
    rules: { required: true, 
    validate: (val) => {
          const validationFunc = handleValidationErrors(2, null, {
            defaultMessage: "Please enter your first name",
            1: "First name should be at least 2 characters",
          });

          const errorMsg = validationFunc(val);

          if (errorMsg) {
            setErrorMessages({ ...errorMessages, firstName: errorMsg });
            return false;
          } else {
            return true;
          }
        },
      },
    }
 },
    defaultValue: "",
  });


return <>
<TextInput
           label={required ? label + " *" : label}
          value={value}
          onChangeText={(text) => {
            if (invalid) formControls.clearErrors(name);
            onChange(text);
          }}
          onBlur={() => formControls.trigger(name)}
          />
          {!!invalid && <InputError msg={errorMessages.firstName} /> }
</>


```

This works, sorta (it doesn't seem to run the validation function when the field is empty, as the `required` rule might pick it up first, and that means the defaultMessage doesn't get updated properly.

I figure there must be a better, more RHF-y way to do this, because this feels majorly inefficient. Any clues @bluebill1049 ?

**Comment by bluebill1049** [comment, marked as answer]
hey @jvgeee, why not leverage the return from each attribute?

https://react-hook-form.com/api/useform/register

```tsx
{
  required: 'This is required',
  validate: (a) => a === 'test' && 'Some error message',
  min: {
    value: 2,
    message: 'Min value is 2'
  }
}

// usage:
errors?.required?.message
errors?.valdiate?.message
errors?.min?.message
```

**Reply by jvgeee** [reply]
Holy shit, you are amazing. Thanks for the quick and insightful response, as always.

**Reply by bluebill1049** [reply]
np 👍

**Reply by jvgeee** [reply]
Can I just clarify something @bluebill1049 , with the `validate` functions, doesn't truthy === pass and falsy === fail? 

I thought if I did something like:

```
validate: (a) => a > 3
```

where a === 5, then this would validate with no errors?

If that's the case, wouldn't ` validate: (a) => a === 'test' && 'Some error message'` trigger `invalid` on `useController` when a === 'test'?

**Reply by bluebill1049** [reply]
only `boolean` `true` is valid. you can keep it simpler with `if` and `else` statements. I did the above quickly.

**Reply by jvgeee** [reply]
Got it. So if my validate function returns `true` then it's valid, if it returns `false` or a message then it's invalid and will trigger. Makes sense, thanks mate!
